### PR TITLE
[GA Skills SDK][Virtual Assistant] Update MainDialog

### PR DESCRIPTION
**Pull request by VictorGrycuk**
### Purpose
*What is the context of this pull request? Why is it being done?*

### Changes
*Are there any changes that need to be called out as significant or particularly difficult to grasp? (Include illustrative screenshots for context if applicable.)*

### Tests
*Is this covered by existing tests or new ones? If no, why not?*

### Feature Plan
*Are there any remaining steps or dependencies before this issue can be fully resolved? If so, describe and link to any relevant pull requests or issues.*

#### General
- [ ] I have commented my code, particularly in hard-to-understand areas	
- [ ] I have added or updated the appropriate tests	
- [ ] I have updated related documentation

### CODE REVIEW

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 1]
```diff
@@ -13,78 +12,125 @@ import { LuisRecognizer, QnAMakerResult, QnAMaker } from 'botbuilder-ai';
...
+        this.addDialog(new TextPrompt(TextPrompt.name));
+        this.initialDialogId = MainDialog.name;
+
         // Register dialogs
-        this.onboardingDialog = onboardingDialog
-        this.switchSkillDialog = switchSkillDialog
+        this.onboardingDialog = onboardingDialog;
+        this.switchSkillDialog = switchSkillDialog;
         this.addDialog(this.onboardingDialog);
         this.addDialog(this.switchSkillDialog);
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: I think that `QnAMakerDialog` was added in botbuilder-js, can we validate this?

https://github.com/microsoft/botbuilder-js/blob/master/libraries/botbuilder-ai/src/qnaMakerDialog.ts#L55

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 2]
```diff
@@ -13,78 +12,125 @@ import { LuisRecognizer, QnAMakerResult, QnAMaker } from 'botbuilder-ai';
...
+                // Run LUIS recognition on General model and store result in turn state.
+                const generalLuis: LuisRecognizer | undefined = localizedServices.luisServices.get("general");
+                if (generalLuis !== undefined) {
+                    const generalResult: RecognizerResult = await generalLuis.recognize(innerDc.context);
+                    innerDc.context.turnState.set(StateProperties.GeneralResult, generalResult);
+                }
+            }
+
+            // Check for any interruptions
+            const interrupted: Boolean = await this.interruptDialog(innerDc);
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
            const interrupted: boolean = await this.interruptDialog(innerDc);
```

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 3]
```diff
@@ -13,78 +12,125 @@ import { LuisRecognizer, QnAMakerResult, QnAMaker } from 'botbuilder-ai';
...
+
+            if (interrupted !== undefined) {
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
            if (interrupted) {
```

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 4]
```diff
@@ -93,14 +139,22 @@ export class MainDialog extends ActivityHandlerDialog {
...
                 const generalLuis: LuisRecognizer | undefined = localizedServices.luisServices.get("general");
                 if (generalLuis !== undefined) {
                     const generalResult: RecognizerResult = await generalLuis.recognize(innerDc.context);
-                    innerDc.context.turnState.set(StateProperties.generalResult, generalResult);
+                    innerDc.context.turnState.set(StateProperties.GeneralResult, generalResult);
                 }
             }
+
+            // Check for any interruptions
+            const interrupted: Boolean = await this.interruptDialog(innerDc);
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
            const interrupted: boolean = await this.interruptDialog(innerDc);
```

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 5]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
-        const activity: Activity = dc.context.activity;
-        const userProfile: IUserProfileState = await this.userProfileState.get(dc.context, { name: '' });
-        const dialog: Dialog | undefined = dc.activeDialog !== undefined ? dc.findDialog(dc.activeDialog.id) : undefined;
+    private async interruptDialog(innerDc: DialogContext): Promise<Boolean> {
+        let interrupted: boolean = false;
+        const activity: Activity = innerDc.context.activity;
+        const userProfile: IUserProfileState = await this.userProfileState.get(innerDc.context, { name: '' });
+        const dialog: Dialog | undefined = innerDc.activeDialog !== undefined ? innerDc.findDialog(innerDc.activeDialog.id) : undefined;
 
         if (activity.type === ActivityTypes.Message && activity.text.trim().length > 0) {
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
        if (activity.type === ActivityTypes.Message && activity.text !== undefined && activity.text.trim().length > 0) {
```

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 6]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
-             // Check if we need to switch skills.
-             if (isSkill) {
+            // Check if we need to switch skills.
+            if (isSkill !== undefined) {
                 if (dialog !== undefined) {
                     if (intent !== dialog.id && dispatchResult.intents[intent].score > 0.9) {
-                        const identifiedSkill: ISkillManifest | undefined = SkillRouter.isSkill(this.settings.skills, LuisRecognizer.topIntent(dispatchResult).toString());
-     
-                        if (identifiedSkill) {
+                        const identifiedSkill: EnhancedBotFrameworkSkill;
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: This is not initializate as C# [does] (https://github.com/microsoft/botframework-solutions/blob/10ae4d2f0fa595e9d1b1f5d5ba11f15fd3160e9a/templates/csharp/VA/VA/Dialogs/MainDialog.cs#L183)

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 7]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
 
             // Get Dispatch LUIS result from turn state.
-            const dispatchResult: RecognizerResult = dc.context.turnState.get(StateProperties.dispatchResult);
+            const dispatchResult: RecognizerResult = innerDc.context.turnState.get(StateProperties.dispatchResult);
             const intent: string = LuisRecognizer.topIntent(dispatchResult);
             
-             // Check if we need to switch skills.
-             if (isSkill) {
+            // Check if we need to switch skills.
+            if (isSkill !== undefined) {
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
            if (isSkill) {
```

Comment (Batta32) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Also, why is not present the `isSkillIntent(dispatchIntent)` as C#?

https://github.com/microsoft/botframework-solutions/blob/10ae4d2f0fa595e9d1b1f5d5ba11f15fd3160e9a/templates/csharp/VA/VA/Dialogs/MainDialog.cs#L180

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 8]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
+                        if (this.skillsConfig.skills.get(intent)) {
                             const prompt: Partial<Activity> = this.templateEngine.generateActivityForLocale('SkillSwitchPrompt', { skill: identifiedSkill.name });
-                            await dc.beginDialog(this.switchSkillDialog.id, new SwitchSkillDialogOptions(prompt as Activity, identifiedSkill));
-
-                            return InterruptionAction.Waiting;
+                            await innerDc.beginDialog(this.switchSkillDialog.id, new SwitchSkillDialogOptions(prompt as Activity, identifiedSkill));
+                            interrupted = true;
+                        } else {
+                            throw new Error(`${intent} is not in the skills configuration`);
                         }
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: This is not correct here, you should refactor the `if` statement

Check [C#](https://github.com/microsoft/botframework-solutions/blob/10ae4d2f0fa595e9d1b1f5d5ba11f15fd3160e9a/templates/csharp/VA/VA/Dialogs/MainDialog.cs#L189)

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 9]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
-
-                            if(isSkill) {
-                                // If current dialog is a skill, allow it to handle its own help intent.
-                                await dc.continueDialog();
-                                break;
-                            } else {
-                                await dc.context.sendActivity(this.templateEngine.generateActivityForLocale('HelpCard', userProfile));
-                                return InterruptionAction.Resume;
+                            if (!isSkill) {
+                                await innerDc.context.sendActivity(this.templateEngine.generateActivityForLocale('HelpCard', userProfile));
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Missing comment above here

https://github.com/microsoft/botframework-solutions/blob/10ae4d2f0fa595e9d1b1f5d5ba11f15fd3160e9a/templates/csharp/VA/VA/Dialogs/MainDialog.cs#L226

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 10]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
+        return interrupted;
     }
 
-    // Runs when the dialog stack is empty, and a new member is added to the conversation. Can be used to send an introduction activity.
-    protected async onMembersAdded(innerDc: DialogContext): Promise<void> {
-        const userProfile: IUserProfileState = await this.userProfileState.get(innerDc.context, { name: '' });
+    private async onBoardingStep(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {
+        const userProfile: IUserProfileState = await this.userProfileState.get(stepContext.context, { name: '' });
 
         if (userProfile === undefined || userProfile.name.trim().length === 0) {
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
        if (userProfile.name === undefined || userProfile.name.trim().length === 0) {
```

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 11]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
-            // Send new user intro card.
-            await innerDc.context.sendActivity(this.templateEngine.generateActivityForLocale('NewUserIntroCard', userProfile));
+            await stepContext.beginDialog(OnboardingDialog.name);
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
            return await stepContext.beginDialog(this.onboardingDialog.id);
```

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 12]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
-            await innerDc.beginDialog(OnboardingDialog.name);
-        } else {
-            // Send returning user intro card.
-            await innerDc.context.sendActivity(this.templateEngine.generateActivityForLocale('ReturningUserIntroCard', userProfile));
-        }
+        return await stepContext.next();
+    }
+    
+    private async introStep(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {
+    // Use the text provided in FinalStepAsync or the default if it is the first time.
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: This comment is not alligned to the code

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 13]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
+        let promptOptions: PromptOptions = {
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
        const promptOptions: PromptOptions = {
```

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 14]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
             // Get dispatch result from turn state.
-            const dispatchResult: RecognizerResult = innerDc.context.turnState.get(StateProperties.dispatchResult);
+            const dispatchResult: RecognizerResult = stepContext.context.turnState.get(StateProperties.dispatchResult);
             const dispatch: string = LuisRecognizer.topIntent(dispatchResult);
 
-            // Check if the dispatch intent maps to a skill.
-            const identifiedSkill: ISkillManifest | undefined = SkillRouter.isSkill(this.settings.skills, dispatch);
-
-            if (identifiedSkill !== undefined) {
+            if (this.isSkillIntent !== undefined) {
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: This line should call to the `isSkillIntent` method, you are not calling this correctly.

Check [this](https://github.com/microsoft/botframework-solutions/blob/10ae4d2f0fa595e9d1b1f5d5ba11f15fd3160e9a/templates/csharp/VA/VA/Dialogs/MainDialog.cs#L323)

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 15]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
+                const dispatchIntentSkill = dispatch.toString();
+                const skillDialogArgs: SkillDialogArgs = skillDialogArgs{ skillId = dispatchIntentSkill };
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
                const skillDialogArgs: SkillDialogArgs = new SkillDialogArgs().skillId = dispatchIntentSkill;
```

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 16]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
+
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Delete this blank space

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 17]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
+                
                 // Start the skill dialog.
-                await innerDc.beginDialog(identifiedSkill.id);
+                return await stepContext.beginDialog(dispatchIntentSkill, skillDialogArgs);
+            
             } else if (dispatch === 'q_faq') {
                 const qnaMaker: QnAMaker | undefined = localizedServices.qnaServices.get('faq');
                 if (qnaMaker !== undefined) {
-                    await this.callQnaMaker(innerDc, qnaMaker);
+                    return await this.callQnaMaker(stepContext, qnaMaker);
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: This piece of code is different to [C#](https://github.com/microsoft/botframework-solutions/blob/10ae4d2f0fa595e9d1b1f5d5ba11f15fd3160e9a/templates/csharp/VA/VA/Dialogs/MainDialog.cs#L331)

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 18]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
-                DialogContextEx.suppressCompletionMessage(innerDc, true);
+            } else if (dispatch === 'q_chitchat') {
+                DialogContextEx.suppressCompletionMessage(stepContext, true);
                 const qnaMaker: QnAMaker | undefined = localizedServices.qnaServices.get('chitchat');
-                if (qnaMaker !== undefined){
-                    await this.callQnaMaker(innerDc, qnaMaker);
+                if (qnaMaker !== undefined) {
+                    return await this.callQnaMaker(stepContext, qnaMaker);
                 }
             } else {
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: This piece of code is different to [C#](https://github.com/microsoft/botframework-solutions/blob/10ae4d2f0fa595e9d1b1f5d5ba11f15fd3160e9a/templates/csharp/VA/VA/Dialogs/MainDialog.cs#L337)

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 19]
```diff
@@ -109,253 +163,187 @@ export class MainDialog extends ActivityHandlerDialog {
...
-            }
-        }
+                DialogContextEx.suppressCompletionMessage(stepContext, true);
+                await stepContext.context.sendActivity(this.templateEngine.generateActivityForLocale('UnsupportedMessage', userProfile));
+                
+                return await stepContext.next();
+            } 
+        } 
+
+        return await stepContext.next();
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: This should be inside of an `else` statement of the if which checks if the `activity.text` is undefined or empty.

https://github.com/microsoft/botframework-solutions/blob/10ae4d2f0fa595e9d1b1f5d5ba11f15fd3160e9a/templates/csharp/VA/VA/Dialogs/MainDialog.cs#L337

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 20]
```diff
@@ -366,20 +354,22 @@ export class MainDialog extends ActivityHandlerDialog {
             await dc.cancelAllDialogs();
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Delete this blank line

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 21]
```diff
@@ -366,20 +354,22 @@ export class MainDialog extends ActivityHandlerDialog {
...
 
-        if (answer !== undefined && answer.length > 0){
+        if (answer !== undefined && answer.length > 0) {
             await innerDc.context.sendActivity(answer[0].answer, answer[0].answer);
         } else {
             await innerDc.context.sendActivity(this.templateEngine.generateActivityForLocale('UnsupportedMessage', userProfile));
         }
+
+        return {status: DialogTurnStatus.empty};
     }
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Can we delete this? This doesn't appear in C# [implementation](https://github.com/microsoft/botframework-solutions/blob/10ae4d2f0fa595e9d1b1f5d5ba11f15fd3160e9a/templates/csharp/VA/VA/Dialogs/MainDialog.cs)

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 22]
```diff
@@ -366,20 +354,22 @@ export class MainDialog extends ActivityHandlerDialog {
...
     
     private async storeOutgoingActivities(turnContext: TurnContext, activities: Partial<Activity>[], next: () => Promise<ResourceResponse[]>): Promise<ResourceResponse[]> {
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Can we change the `==` to `===`?

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 23]
```diff
@@ -398,17 +388,12 @@ export class MainDialog extends ActivityHandlerDialog {
 
         return await next();
     }
-}
 
-enum Events {
-    location = 'VA.Location',
-    timeZone = 'VA.Timezone'
-}
+    private isSkillIntent(dispatchIntent: string): Boolean {
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
    private isSkillIntent(dispatchIntent: string): boolean {
```

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 24]
```diff
@@ -398,17 +388,12 @@ export class MainDialog extends ActivityHandlerDialog {
...
+        if (dispatchIntent.toLowerCase() === 'l_general' || dispatchIntent.toLowerCase() === 'q_chichat' || dispatchIntent.toLowerCase() === 'q_faq' || dispatchIntent.toLowerCase() === 'none') {
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: This line is too long.. can we put one under another?

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 25]
```diff
@@ -2,89 +2,154 @@
...
-        this.switchSkillDialog = switchSkillDialog
-        this.addDialog(this.onboardingDialog);
+        this.onBoardingDialog = onBoardingDialog;
+        this.switchSkillDialog = switchSkillDialog;
+        this.addDialog(this.onBoardingDialog);
         this.addDialog(this.switchSkillDialog);
 
+        // Register a QnAMakerDialog for each registered knowledgebase and ensure localised responses are provided.
+        const localizedServices: ICognitiveModelSet  = this.services.getCognitiveModels();
+        localizedServices.qnaConfiguration.forEach( (knowledgebase: QnAMakerDialog): void => {
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
        localizedServices.qnaConfiguration.forEach((knowledgebase: QnAMakerDialog): void => {
```

This was marked by ESLint?

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 26]
```diff
@@ -2,89 +2,154 @@
...
+        const qnaDialog = new QnAMakerDialog(
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
        const qnaDialog: QnAMakerDialog = new QnAMakerDialog(
```

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 27]
```diff
@@ -2,89 +2,154 @@
...
+            knowledgeBaseId = knowledgebase.value.knowledgeBaseId,
+            endpointKey = knowledgebase.value.endpointKey,
+            hostName = knowledgebase.value.host,
+            noAnswer = this.templateEngine.generateActivityForLocale("UnsupportedMessage"),
+            activeLearningCardTitle = this.templateEngine.generateActivityForLocale("QnaMakerAdaptiveLearningCardTitle"),
+            cardNoMatchText = this.templateEngine.generateActivityForLocale("QnaMakerNoMatchText"))
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: It's missing the `.text` as [C#](https://github.com/microsoft/botframework-solutions/blob/10ae4d2f0fa595e9d1b1f5d5ba11f15fd3160e9a/templates/csharp/VA/VA/Dialogs/MainDialog.cs#L81)

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 28]
```diff
@@ -2,89 +2,154 @@
...
+                    innerDc.context.turnState.set(StateProperties.GeneralResult, generalResult);
+                }
+            }
+
+            // Check for any interruptions
+            const interrupted: boolean = await this.interruptDialog(innerDc);
+
+            if (interrupted) {
+                // If dialog was interrupted, return EndOfTurn
+                return MainDialog.EndOfTurn;
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Is `EndOfTurn` a class variable?

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 29]
```diff
@@ -109,253 +182,183 @@ export class MainDialog extends ActivityHandlerDialog {
...
+                    const identifiedSkill: EnhancedBotFrameworkSkill | undefined = this.skillsConfig.skills.get(intent);
+                    if (identifiedSkill !== undefined) {
+                        const prompt: Partial<Activity> = this.templateEngine.generateActivityForLocale('SkillSwitchPrompt', { skill: identifiedSkill.name });
+                        await innerDc.beginDialog(this.switchSkillDialog.id, new SwitchSkillDialogOptions(prompt as Activity, identifiedSkill));
+                        interrupted = true;
+                    } else {
+                        throw new Error(`${intent} is not in the skills configuration`);
                     }
-                }
-            }
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Delete this blank space

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 30]
```diff
@@ -109,253 +182,183 @@ export class MainDialog extends ActivityHandlerDialog {
...
-                 if(generalResult.intents[intent].score > 0.5) {
-                    switch(intent.toString()) {
+                 if (generalResult.intents[intent].score > 0.5) {
+                    switch(intent) {
                         case 'Cancel': { 
-                            // Suppress completion message for utility functions.
-                            DialogContextEx.suppressCompletionMessage(dc, true);
-
-                            await dc.context.sendActivity(this.templateEngine.generateActivityForLocale('CancelledMessage', userProfile));
-                            await dc.cancelAllDialogs();
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Delete this blank space

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 31]
```diff
@@ -109,253 +182,183 @@ export class MainDialog extends ActivityHandlerDialog {
...
 
-        return InterruptionAction.NoAction;
+        return interrupted;
     }
 
-    // Runs when the dialog stack is empty, and a new member is added to the conversation. Can be used to send an introduction activity.
-    protected async onMembersAdded(innerDc: DialogContext): Promise<void> {
-        const userProfile: IUserProfileState = await this.userProfileState.get(innerDc.context, { name: '' });
+    private async onBoardingStep(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {
+        const userProfile: IUserProfileState = await this.userProfileState.get(stepContext.context, { name: '' });
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Delete blank space

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 32]
```diff
@@ -109,253 +182,183 @@ export class MainDialog extends ActivityHandlerDialog {
...
-            const dispatchResult: RecognizerResult = innerDc.context.turnState.get(StateProperties.dispatchResult);
+            const dispatchResult: RecognizerResult = stepContext.context.turnState.get(StateProperties.dispatchResult);
             const dispatch: string = LuisRecognizer.topIntent(dispatchResult);
-
-            // Check if the dispatch intent maps to a skill.
-            const identifiedSkill: ISkillManifest | undefined = SkillRouter.isSkill(this.settings.skills, dispatch);
-
-            if (identifiedSkill !== undefined) {
+            if (this.isSkillIntent(dispatch)) {
+                const dispatchIntentSkill = dispatch.toString();
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: `dispatch` is already a string, the `toString()` is not necessary

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 33]
```diff
@@ -364,51 +367,37 @@ export class MainDialog extends ActivityHandlerDialog {
...
-    timeZone = 'VA.Timezone'
-}
+    private isSkillIntent(dispatchIntent: string): boolean {
+        if (dispatchIntent.toLowerCase() === 'l_general' || 
+        dispatchIntent.toLowerCase() === 'q_chichat' || 
+        dispatchIntent.toLowerCase() === 'q_faq' || 
+        dispatchIntent.toLowerCase() === 'none') {
+
+            return false;
+        }
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
    private isSkillIntent(dispatchIntent: string): boolean {
        if (dispatchIntent.toLowerCase() === 'l_general' || 
            dispatchIntent.toLowerCase() === 'q_chichat' || 
            dispatchIntent.toLowerCase() === 'q_faq' || 
            dispatchIntent.toLowerCase() === 'none') {

            return false;
        }
```

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts (hunk 25 above)
Comment (DiegoCardozo94) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Yes, qnaConfiguration is not present in ICognitiveModelSet interface

Comment (Batta32) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: I was asking about the space that I removed 😂

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts (hunk 28 above)
Comment (DiegoCardozo94) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Sorry for ask, but what do you mean here?

Comment (Batta32) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: You are using `MainDialog.EndOfTurn` which means that `EndOfTurn` is a property defined as class variable (you can't access to that using by the object created (using `this`).

What I asked is if  `EndOfTurn` is inherited and defined as `static` or not.

Comment (DiegoCardozo94) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: Yes, `EndOfTurn` is static.

Comment (Batta32) [reply templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]: ACK, so it's okay to be like it's now 😊

File: templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts [hunk 34]
```diff
@@ -83,7 +83,7 @@ export class MainDialog extends ComponentDialog {
         ];
 
         this.addDialog(new WaterfallDialog (MainDialog.name, steps));
-        this.addDialog(new TextPrompt(TextPrompt.name));
+        this.addDialog(new TextPrompt(TextPrompt.name))
```
Comment (Batta32) [inline templates/typescript/samples/sample-assistant/src/dialogs/mainDialog.ts]:
```suggestion
        this.addDialog(new TextPrompt(TextPrompt.name));
```

**Pull request merged**